### PR TITLE
[WirelengthAnalyzer] Add new connectivity rules

### PR DESCRIPTION
**Pull request by zakn-amd**
Add connectivity rules for cells of type:
- `MMCE4_ADV`
- `GTYE4_CHANNEL`
- `GTYE4_COMMON`
- `PCIE40E4`
- `URAM288`
- `STARTUPE3`
- `ICAPE3`

Also add a new PIP wirelength entry, some small refactors, and documentation updates to reflect above changes.

### CODE REVIEW

File: docs/score.md [hunk 1]
```diff
@@ -114,12 +114,17 @@ BEL input and output pins for each type of PhysCell.
...
 |`MUXF7`, `MUXF8`, `MUXF9`                                                                 | (all) <- (all)                | Intrasite Mux      |
 |`IBUFCTRL`                                                                                | (all) <- (all)                | Input Buffer       |
 |`DSP_A_B_DATA`, `DSP_C_DATA`, `DSP_M_DATA`,<br>`DSP_PREADD_DATA`, `DSP_OUTPUT`, `DSP_ALU` | (none) <- (none) [see note](#dsp-cell-connectivity) | DSP Logic |
 |`DSP_MULTIPLIER`, `DSP_PREADD`                                                            | (all) <- (all) [see note](#dsp-cell-connectivity) | DSP Logic |
+|`PCIE40E4`                                                                                | (none) <- (none)              | PCIe hard macro    |
+|`GTYE4_CHANNEL`,`GTYE4_COMMON`                                                            | (none) <- (none)              | gigabit transceiver components |
+|`STARTUPE3`,`ICAPE3`                                                                      | (none) <- (none)              | Device configuration components |
```
Comment (eddieh-xlnx) [inline docs/score.md:127]:
```suggestion
|`PCIE40E4`                                                                                | (none) <- (none)              | PCIe Hard Macro    |
|`GTYE4_CHANNEL`,`GTYE4_COMMON`                                                            | (none) <- (none)              | Gigabit Transceiver Components |
|`STARTUPE3`,`ICAPE3`                                                                      | (none) <- (none)              | Device Configuration Components |
```

File: wirelength_analyzer/wa.py [hunk 2]
```diff
@@ -355,7 +354,7 @@ def add_all_nets_to_graph(self):
                 # Omit source (BELPins) that don't have any fanout
                 if len(branch.branches) == 0:
                     continue
-                if sl[branch.routeSegment.belPin.bel] in ('BUFCE'):
+                if sl[branch.routeSegment.belPin.bel] in ('BUFCE', 'BUFG_GT', 'BUFG_GT_SYNC'):
```
Comment (eddieh-xlnx) [inline wirelength_analyzer/wa.py:357]: Should this be moved into device data too?

File: wirelength_analyzer/xcvup_device_data.py [hunk 3]
```diff
@@ -97,6 +108,7 @@ def __contains__(self, item):
             (re.compile(r'INT_NODE_GLOBAL_\d{1,2}_INT_OUT[01]'),     0),
             (re.compile(r'IMUX_[EW]\d{1,2}'),                        0),
             (re.compile(r'IMUX_(CMT_)?XIPHY\d{1,2}'),                0),
+            (re.compile(r'IMUXOUT\d{1,2}'),                          0),
```
Comment (eddieh-xlnx) [inline wirelength_analyzer/xcvup_device_data.py:111]: Can you give an example of an IMUXOUT pip?

File: wirelength_analyzer/xcvup_device_data.py [hunk 4]
```diff
@@ -53,6 +53,17 @@ def __contains__(self, item):
             'RAMB36E2':        self.default_sequential,
             'RAMB18E2':        self.default_sequential,
 
+            'MMCME4_ADV':      self.default_sequential,
```
Comment (eddieh-xlnx) [inline wirelength_analyzer/xcvup_device_data.py:56]: I think `default_sequential` is no longer appropriate for some of these new blocks. Perhaps we should remain this and `default_combinatorial` to `none_to_none` and `all_to_all` respectively to match the webpage?

File: docs/score.md [hunk 5]
```diff
@@ -114,12 +114,17 @@ BEL input and output pins for each type of PhysCell.
 |`RAMD32`, `RAMS32`                                                                        | `O5` and `O6` <- `A0` to `A4` | Distributed Memory |
 |`RAMD64E`, `RAMS64E`                                                                      | `O6` <- `A0` to `A5`          | Distributed Memory |
 |`RAMB36E2`, `RAMB18E2`                                                                    | (none) <- (none)              | Block Memory       |
+|`URAM288`                                                                                 | (none) <- (none)              | UltraRAM           |
```
Comment (eddieh-xlnx) [inline docs/score.md:117]:
```suggestion
|`URAM288`                                                                                 | (none) <- (none)              | UltraRAM Memory    |
```

File: wirelength_analyzer/xcvup_device_data.py (hunk 3 above)
Comment (zakn-amd) [reply wirelength_analyzer/xcvup_device_data.py:111]: see the switchbox in the tile `INT_INTF_L_PCIE4_X100Y7`

File: wirelength_analyzer/wa.py (hunk 2 above)
Comment (zakn-amd) [reply wirelength_analyzer/wa.py:357]: yes I think that would make sense, I'll change it

Comment (zakn-amd) [reply wirelength_analyzer/wa.py]: https://github.com/Xilinx/fpga24_routing_contest/pull/40/commits/e7ac25f588ea425c97920bfa9c2d358f6aa1f5f3

File: wirelength_analyzer/xcvup_device_data.py (hunk 3 above)
Comment (eddieh-xlnx) [reply wirelength_analyzer/xcvup_device_data.py:111]: One example is `INT_INTF_L_PCIE4_X100Y7/INT_INTF_L_PCIE4.DELAYWIRE_PCIE->IMUXOUT0`.

File: wirelength_analyzer/xcvup_device_data.py (hunk 4 above)
Comment (zakn-amd) [reply wirelength_analyzer/xcvup_device_data.py]: renamed here: https://github.com/Xilinx/fpga24_routing_contest/pull/40/commits/9c8a3c73adbcc5c7ccd86db82d51d55c7d119182

**Review by eddieh-xlnx — APPROVED** [review]

**Pull request merged**